File: 10/gemini-3.0-pro/go/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	content, err := ioutil.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}

	lines := strings.Split(string(content), "\n")
	var totalPressesPart1 int64 = 0
	var totalPressesPart2 int64 = 0

	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		totalPressesPart1 += solvePart1(line)
		totalPressesPart2 += solvePart2(line)
	}

	fmt.Printf("Day 10 Part 1: %d\n", totalPressesPart1)
	fmt.Printf("Day 10 Part 2: %d\n", totalPressesPart2)
}

func solvePart1(line string) int64 {
	reTarget := regexp.MustCompile(`\[([.#]+)\]`)
	match := reTarget.FindStringSubmatch(line)
	if match == nil {
		return 0
	}
	targetStr := match[1]
	L := len(targetStr)
	target := make([]int, L)
	for i, c := range targetStr {
		if c == '#' {
			target[i] = 1
		} else {
			target[i] = 0
		}
	}

	reButtons := regexp.MustCompile(`\(([\d,]+)\)`)
	buttonMatches := reButtons.FindAllStringSubmatch(line, -1)
	var buttons [][]int
	for _, bm := range buttonMatches {
		b := make([]int, L)
		parts := strings.Split(bm[1], ",")
		for _, p := range parts {
			idx, err := strconv.Atoi(p)
			if err == nil && idx < L {
				b[idx] = 1
			}
		}
		buttons = append(buttons, b)
	}

	B := len(buttons)
	M := make([][]int, L)
	for i := range M {
		M[i] = make([]int, B+1)
	}

	for r := 0; r < L; r++ {
		for c := 0; c < B; c++ {
			M[r][c] = buttons[c][r]
		}
		M[r][B] = target[r]
	}

	pivotRow := 0
	var pivotCols []int
	colToPivotRow := make([]int, B)
	for i := 0; i < B; i++ {
		colToPivotRow[i] = -1
	}

	for c := 0; c < B && pivotRow < L; c++ {
		sel := -1
		for r := pivotRow; r < L; r++ {
			if M[r][c] == 1 {
				sel = r
				break
			}
		}

		if sel == -1 {
			continue
		}

		// Swap
		M[pivotRow], M[sel] = M[sel], M[pivotRow]

		// Eliminate
		for r := 0; r < L; r++ {
			if r != pivotRow && M[r][c] == 1 {
				for k := c; k <= B; k++ {
					M[r][k] ^= M[pivotRow][k]
				}
			}
		}

		pivotCols = append(pivotCols, c)
		colToPivotRow[c] = pivotRow
		pivotRow++
	}

	for r := pivotRow; r < L; r++ {
		if M[r][B] == 1 {
			return 0
		}
	}

	var freeCols []int
	pivotSet := make(map[int]bool)
	for _, p := range pivotCols {
		pivotSet[p] = true
	}
	for c := 0; c < B; c++ {
		if !pivotSet[c] {
			freeCols = append(freeCols, c)
		}
	}

	minPresses := int64(math.MaxInt64)
	numFree := len(freeCols)
	combinations := int64(1) << numFree

	for i := int64(0); i < combinations; i++ {
		x := make([]int, B)
		var currentPresses int64 = 0

		for k := 0; k < numFree; k++ {
			if (i & (1 << k)) != 0 {
				x[freeCols[k]] = 1
				currentPresses++
			}
		}

		for _, pCol := range pivotCols {
			r := colToPivotRow[pCol]
			val := M[r][B]
			for _, fCol := range freeCols {
				if M[r][fCol] == 1 && x[fCol] == 1 {
					val ^= 1
				}
			}
			x[pCol] = val
			if val == 1 {
				currentPresses++
			}
		}

		if currentPresses < minPresses {
			minPresses = currentPresses
		}
	}

	return minPresses
}

func solvePart2(line string) int64 {
	reTarget := regexp.MustCompile(`\{([\d,]+)\}`)
	match := reTarget.FindStringSubmatch(line)
	if match == nil {
		return 0
	}
	targetParts := strings.Split(match[1], ",")
	L := len(targetParts)
	b := make([]float64, L)
	for i, p := range targetParts {
		val, _ := strconv.ParseFloat(p, 64)
		b[i] = val
	}

	reButtons := regexp.MustCompile(`\(([\d,]+)\)`)
	buttonMatches := reButtons.FindAllStringSubmatch(line, -1)
	var A_cols [][]float64
	for _, bm := range buttonMatches {
		col := make([]float64, L)
		parts := strings.Split(bm[1], ",")
		for _, p := range parts {
			idx, err := strconv.Atoi(p)
			if err == nil && idx < L {
				col[idx] = 1.0
			}
		}
		A_cols = append(A_cols, col)
	}

	B := len(A_cols)
	matrix := make([][]float64, L)
	for i := range matrix {
		matrix[i] = make([]float64, B+1)
	}

	for r := 0; r < L; r++ {
		for c := 0; c < B; c++ {
			matrix[r][c] = A_cols[c][r]
		}
		matrix[r][B] = b[r]
	}

	pivotRow := 0
	var pivotCols []int
	colToPivotRow := make([]int, B)
	for i := 0; i < B; i++ {
		colToPivotRow[i] = -1
	}

	for c := 0; c < B && pivotRow < L; c++ {
		sel := -1
		for r := pivotRow; r < L; r++ {
			if math.Abs(matrix[r][c]) > 1e-9 {
				sel = r
				break
			}
		}

		if sel == -1 {
			continue
		}

		matrix[pivotRow], matrix[sel] = matrix[sel], matrix[pivotRow]

		pivotVal := matrix[pivotRow][c]
		for k := c; k <= B; k++ {
			matrix[pivotRow][k] /= pivotVal
		}

		for r := 0; r < L; r++ {
			if r != pivotRow && math.Abs(matrix[r][c]) > 1e-9 {
				factor := matrix[r][c]
				for k := c; k <= B; k++ {
					matrix[r][k] -= factor * matrix[pivotRow][k]
				}
			}
		}

		pivotCols = append(pivotCols, c)
		colToPivotRow[c] = pivotRow
		pivotRow++
	}

	for r := pivotRow; r < L; r++ {
		if math.Abs(matrix[r][B]) > 1e-9 {
			return 0
		}
	}

	var freeCols []int
	pivotSet := make(map[int]bool)
	for _, p := range pivotCols {
		pivotSet[p] = true
	}
	for c := 0; c < B; c++ {
		if !pivotSet[c] {
			freeCols = append(freeCols, c)
		}
	}

	UBs := make([]int64, B)
	for c := 0; c < B; c++ {
		minUb := int64(math.MaxInt64)
		bounded := false
		for r := 0; r < L; r++ {
			if A_cols[c][r] > 0.5 {
				val := int64(b[r] / A_cols[c][r])
				if val < minUb {
					minUb = val
				}
				bounded = true
			}
		}
		if bounded {
			UBs[c] = minUb
		} else {
			UBs[c] = 0
		}
	}

	minTotalPresses := int64(math.MaxInt64)
	found := false

	var search func(freeIdx int, currentFreeVals []int64)
	search = func(freeIdx int, currentFreeVals []int64) {
		if freeIdx == len(freeCols) {
			var currentPresses int64 = 0
			for _, v := range currentFreeVals {
				currentPresses += v
			}

			possible := true
			for _, pCol := range pivotCols {
				r := colToPivotRow[pCol]
				val := matrix[r][B]
				for j, fCol := range freeCols {
					val -= matrix[r][fCol] * float64(currentFreeVals[j])
				}

				if val < -1e-9 {
					possible = false
					break
				}
				longVal := int64(math.Round(val))
				if math.Abs(val - float64(longVal)) > 1e-9 {
					possible = false
					break
				}
				currentPresses += longVal
			}

			if possible {
				if currentPresses < minTotalPresses {
					minTotalPresses = currentPresses
					found = true
				}
			}
			return
		}

		fColIdx := freeCols[freeIdx]
		limit := UBs[fColIdx]

		for val := int64(0); val <= limit; val++ {
			currentFreeVals[freeIdx] = val
			search(freeIdx+1, currentFreeVals)
		}
	}

	search(0, make([]int64, len(freeCols)))

	if found {
		return minTotalPresses
	}
	return 0
}

File: 10/gemini-3.0-pro/rust/src/main.rs
use std::fs;
use regex::Regex;

fn main() {
    let input = fs::read_to_string("input.txt").expect("Unable to read file");
    let mut total_presses: i64 = 0;
    let mut total_presses_part2: i64 = 0;

    for line in input.lines() {
        if line.trim().is_empty() {
            continue;
        }
        total_presses += solve_machine(line);
        total_presses_part2 += solve_part2(line);
    }

    println!("Day 10 Part 1: {}", total_presses);
    println!("Day 10 Part 2: {}", total_presses_part2);
}

fn solve_machine(line: &str) -> i64 {
    let re_target = Regex::new(r"\[([.#]+)\]").unwrap();
    let cap = re_target.captures(line);
    if cap.is_none() {
        return 0;
    }
    let target_str = &cap.unwrap()[1];
    let l = target_str.len();
    let mut target = vec![0; l];
    for (i, c) in target_str.chars().enumerate() {
        if c == '#' {
            target[i] = 1;
        }
    }

    let re_buttons = Regex::new(r"\(([\d,]+)\)").unwrap();
    let mut buttons = Vec::new();
    for cap in re_buttons.captures_iter(line) {
        let mut b = vec![0; l];
        let parts: Vec<&str> = cap[1].split(',').collect();
        for p in parts {
            if let Ok(idx) = p.parse::<usize>() {
                if idx < l {
                    b[idx] = 1;
                }
            }
        }
        buttons.push(b);
    }

    let b_count = buttons.len();
    let mut m = vec![vec![0; b_count + 1]; l];
    for r in 0..l {
        for c in 0..b_count {
            m[r][c] = buttons[c][r];
        }
        m[r][b_count] = target[r];
    }

    let mut pivot_row = 0;
    let mut pivot_cols = Vec::new();
    let mut col_to_pivot_row = vec![-1; b_count];

    for c in 0..b_count {
        if pivot_row >= l {
            break;
        }

        let mut sel = -1;
        for r in pivot_row..l {
            if m[r][c] == 1 {
                sel = r as i32;
                break;
            }
        }

        if sel == -1 {
            continue;
        }

        let sel = sel as usize;
        m.swap(pivot_row, sel);

        for r in 0..l {
            if r != pivot_row && m[r][c] == 1 {
                for k in c..=b_count {
                    m[r][k] ^= m[pivot_row][k];
                }
            }
        }

        pivot_cols.push(c);
        col_to_pivot_row[c] = pivot_row as i32;
        pivot_row += 1;
    }

    for r in pivot_row..l {
        if m[r][b_count] == 1 {
            return 0;
        }
    }

    let mut free_cols = Vec::new();
    for c in 0..b_count {
        if !pivot_cols.contains(&c) {
            free_cols.push(c);
        }
    }

    let mut min_presses = i64::MAX;
    let num_free = free_cols.len();
    let combinations = 1 << num_free;

    for i in 0..combinations {
        let mut x = vec![0; b_count];
        let mut current_presses = 0;

        for k in 0..num_free {
            if (i & (1 << k)) != 0 {
                x[free_cols[k]] = 1;
                current_presses += 1;
            }
        }

        for &p_col in &pivot_cols {
            let r = col_to_pivot_row[p_col] as usize;
            let mut val = m[r][b_count];
            for &f_col in &free_cols {
                if m[r][f_col] == 1 && x[f_col] == 1 {
                    val ^= 1;
                }
            }
            x[p_col] = val;
            if val == 1 {
                current_presses += 1;
            }
        }

        if current_presses < min_presses {
            min_presses = current_presses;
        }
    }

    min_presses
}

fn solve_part2(line: &str) -> i64 {
    let re_target = Regex::new(r"\{([\d,]+)\}").unwrap();
    let cap = re_target.captures(line);
    if cap.is_none() {
        return 0;
    }
    let target_str = &cap.unwrap()[1];
    let target: Vec<f64> = target_str.split(',').map(|s| s.parse().unwrap()).collect();
    let l = target.len();

    let re_buttons = Regex::new(r"\(([\d,]+)\)").unwrap();
    let mut a_cols = Vec::new();
    for cap in re_buttons.captures_iter(line) {
        let mut col = vec![0.0; l];
        let parts: Vec<&str> = cap[1].split(',').collect();
        for p in parts {
            if let Ok(idx) = p.parse::<usize>() {
                if idx < l {
                    col[idx] = 1.0;
                }
            }
        }
        a_cols.push(col);
    }

    let b_count = a_cols.len();
    let mut matrix = vec![vec![0.0; b_count + 1]; l];
    for r in 0..l {
        for c in 0..b_count {
            matrix[r][c] = a_cols[c][r];
        }
        matrix[r][b_count] = target[r];
    }

    let mut pivot_row = 0;
    let mut pivot_cols = Vec::new();
    let mut col_to_pivot_row = vec![-1; b_count];

    for c in 0..b_count {
        if pivot_row >= l { break; }

        let mut sel = -1;
        for r in pivot_row..l {
            if matrix[r][c].abs() > 1e-9 {
                sel = r as i32;
                break;
            }
        }

        if sel == -1 { continue; }

        let sel = sel as usize;
        matrix.swap(pivot_row, sel);

        let pivot_val = matrix[pivot_row][c];
        for k in c..=b_count {
            matrix[pivot_row][k] /= pivot_val;
        }

        for r in 0..l {
            if r != pivot_row && matrix[r][c].abs() > 1e-9 {
                let factor = matrix[r][c];
                for k in c..=b_count {
                    matrix[r][k] -= factor * matrix[pivot_row][k];
                }
            }
        }

        pivot_cols.push(c);
        col_to_pivot_row[c] = pivot_row as i32;
        pivot_row += 1;
    }

    for r in pivot_row..l {
        if matrix[r][b_count].abs() > 1e-9 {
            return 0;
        }
    }

    let mut free_cols = Vec::new();
    for c in 0..b_count {
        if !pivot_cols.contains(&c) {
            free_cols.push(c);
        }
    }

    let mut ubs = vec![0; b_count];
    for c in 0..b_count {
        let mut min_ub = i64::MAX;
        let mut bounded = false;
        for r in 0..l {
            if a_cols[c][r] > 0.5 {
                let val = (target[r] / a_cols[c][r]).floor() as i64;
                if val < min_ub {
                    min_ub = val;
                }
                bounded = true;
            }
        }
        ubs[c] = if bounded { min_ub } else { 0 };
    }

    let mut min_total_presses = i64::MAX;
    let mut found = false;
    let mut current_free_vals = vec![0; free_cols.len()];

    search(0, &free_cols, &mut current_free_vals, &ubs, &pivot_cols, &col_to_pivot_row, &matrix, &mut min_total_presses, &mut found);

    if found { min_total_presses } else { 0 }
}

fn search(
    free_idx: usize,
    free_cols: &Vec<usize>,
    current_free_vals: &mut Vec<i64>,
    ubs: &Vec<i64>,
    pivot_cols: &Vec<usize>,
    col_to_pivot_row: &Vec<i32>,
    matrix: &Vec<Vec<f64>>,
    min_total_presses: &mut i64,
    found: &mut bool
) {
    if free_idx == free_cols.len() {
        let mut current_presses: i64 = current_free_vals.iter().sum();
        let mut possible = true;

        for &p_col in pivot_cols {
            let r = col_to_pivot_row[p_col] as usize;
            let mut val = matrix[r][matrix[0].len() - 1];
            for (j, &f_col) in free_cols.iter().enumerate() {
                val -= matrix[r][f_col] * (current_free_vals[j] as f64);
            }

            if val < -1e-9 { possible = false; break; }
            let long_val = val.round() as i64;
            if (val - (long_val as f64)).abs() > 1e-9 { possible = false; break; }
            current_presses += long_val;
        }

        if possible {
            if current_presses < *min_total_presses {
                *min_total_presses = current_presses;
                *found = true;
            }
        }
        return;
    }

    let f_col_idx = free_cols[free_idx];
    let limit = ubs[f_col_idx];

    for val in 0..=limit {
        current_free_vals[free_idx] = val;
        search(free_idx + 1, free_cols, current_free_vals, ubs, pivot_cols, col_to_pivot_row, matrix, min_total_presses, found);
    }
}
